src/ShallowWrapper.lua

```diff
@@ -0,0 +1,228 @@
+local Children = require(script.Parent.PropMarkers.Children)
+local ElementKind = require(script.Parent.ElementKind)
+local ElementUtils = require(script.Parent.ElementUtils)
+local snapshot = require(script.Parent.snapshot)
+
+local ShallowWrapper = {}
+local ShallowWrapperMetatable = {
+	__index = ShallowWrapper,
+}
+
+local function getTypeFromVirtualNode(virtualNode)
+	local element = virtualNode.currentElement
+	local kind = ElementKind.of(element)
+
+	if kind == ElementKind.Host then
+		return {
+			kind = ElementKind.Host,
+			className = element.component,
+		}
+	elseif kind == ElementKind.Function then
+		return {
+			kind = ElementKind.Function,
+			functionComponent = element.component,
+		}
+	elseif kind == ElementKind.Stateful then
+		return {
+			kind = ElementKind.Stateful,
+			component = element.component,
+		}
+	else
+		error(('shallow wrapper does not support element of kind %q'):format(tostring(kind)))
+	end
+end
+
+local function findNextVirtualNode(virtualNode, maxDepth)
```

Comment:
for my own knowledge, what's the role of this function? If we call Shallow.new why can't we assume we passed in the next virtual node?

Reply:
That is the thing that gives the illusion of shallow rendering, I have not documented it a lot but if you take a look at the notes repo, checkout the tips page there is a example when you are using `depth = 2`.

So `findNextVirtualNode` will get the next virtual node that wraps a host element (a roblox instance). When you allow more _depth_ to the shallow wrapper, it tries to render to roblox instances as much as it can.

I have written a lot of tests to verify the behavior of depth, it's a bit confusing honestly. I've changed the definition a few times before getting this version. It might be something worth reviewing together too if you want

Reply:
Depth might be more intuitive if it still preserves the parent. I tried running some of the tests like stateful-component-children with a depth of 2, and I would have expected CoolComponent to still be in my snapshot, but get to see its children.